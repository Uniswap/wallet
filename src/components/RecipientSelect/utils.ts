import { SectionListData } from 'react-native'
import { SearchableRecipient } from 'src/components/RecipientSelect/types'
import { unique } from 'src/utils/array'

interface HasAddress {
  address: string
}

export function uniqueAddressesOnly<T extends HasAddress>(objectsWithAddress: T[]): T[] {
  // the input array must be objects that have an obj.address field
  // had to cast to any because ts doesn't recognize it as HasAddress... maybe issue with unique
  return unique(
    objectsWithAddress,
    (v, i, a) => a.findIndex((v2) => v2.address === v.address) === i
  )
}

export function filterSections(
  sections: SectionListData<SearchableRecipient>[],
  filteredAddresses: string[]
): { title: string; data: SearchableRecipient[] }[] {
  return sections
    .map((section) => {
      const { title, data } = section
      return { title, data: data.filter((item) => filteredAddresses.includes(item.address)) }
    })
    .filter((section) => section.data.length > 0)
}
