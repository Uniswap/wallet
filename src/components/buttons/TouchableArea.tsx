import { createBox } from '@shopify/restyle'
import { impactAsync, ImpactFeedbackStyle } from 'expo-haptics'
import React, { ComponentProps, PropsWithChildren, useCallback, useMemo } from 'react'
import { GestureResponderEvent, TouchableOpacity, TouchableOpacityProps } from 'react-native'
import {
  Easing,
  useAnimatedStyle,
  useSharedValue,
  withDelay,
  withTiming,
} from 'react-native-reanimated'
import { withAnimated } from 'src/components/animated'
import { TraceEvent } from 'src/components/telemetry/TraceEvent'
import { ReactNativeEvent } from 'src/features/telemetry/constants'
import { TelemetryEventProps } from 'src/features/telemetry/types'
import { defaultHitslopInset } from 'src/styles/sizing'
import { Theme } from 'src/styles/theme'

const TouchableBox = createBox<Theme, TouchableOpacityProps>(TouchableOpacity)
const AnimatedTouchableBox = withAnimated(TouchableBox)

const ScaleTimingConfigIn = { duration: 50, easing: Easing.ease }
const ScaleTimingConfigOut = { duration: 75, easing: Easing.ease }

export type BaseButtonProps = PropsWithChildren<
  ComponentProps<typeof TouchableBox> & {
    hapticFeedback?: boolean
    hapticStyle?: ImpactFeedbackStyle
    scaleTo?: number
  } & TelemetryEventProps
>

/**
 * This component wraps children in a TouchableBox and adds tracking. If you are trying to implement a standard button DO NOT USE this component. Use the Button component instead with the desired size and emphasis.
 * Examples of when to use this are:
 *  - clickable text
 *  - clickable icons (different from an icon button which has a bg color, border radius, and a border)
 *  - custom elements that are clickable (e.g. rows, cards, headers)
 */
export function TouchableArea({
  hapticFeedback = false,
  hapticStyle,
  scaleTo,
  onPress,
  children,
  name: elementName,
  eventName,
  events,
  properties,
  activeOpacity = 0.75,
  ...rest
}: BaseButtonProps): JSX.Element {
  const scale = useSharedValue(1)

  const onPressHandler = useCallback(
    (event: GestureResponderEvent) => {
      if (!onPress) return

      if (hapticFeedback) {
        impactAsync(hapticStyle)
      }

      onPress(event)
    },
    [onPress, hapticFeedback, hapticStyle]
  )

  const onPressInHandler = useMemo(() => {
    if (!scaleTo) return
    return () => {
      scale.value = withTiming(scaleTo, ScaleTimingConfigIn)
    }
  }, [scale, scaleTo])

  const onPressOutHandler = useMemo(() => {
    if (!scaleTo) return
    return () => {
      scale.value = withDelay(50, withTiming(1, ScaleTimingConfigOut))
    }
  }, [scale, scaleTo])

  const baseProps: ComponentProps<typeof TouchableBox> = {
    onPress: onPressHandler,
    onPressIn: onPressInHandler,
    onPressOut: onPressOutHandler,
    activeOpacity,
    hitSlop: defaultHitslopInset,
    testID: elementName,
    ...rest,
  }

  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{ scale: scale.value }],
    }
  })

  if (!eventName) {
    return (
      <AnimatedTouchableBox style={scaleTo ? animatedStyle : undefined} {...baseProps}>
        {children}
      </AnimatedTouchableBox>
    )
  }

  return (
    <TraceEvent
      elementName={elementName}
      eventName={eventName}
      events={[ReactNativeEvent.OnPress, ...(events || [])]}
      properties={properties}>
      <AnimatedTouchableBox {...baseProps} style={scaleTo ? animatedStyle : undefined}>
        {children}
      </AnimatedTouchableBox>
    </TraceEvent>
  )
}

export const AnimatedTouchableArea = withAnimated(TouchableArea)
