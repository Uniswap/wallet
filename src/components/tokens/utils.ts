import { TFunction } from 'i18next'
import { SafetyLevel } from 'src/data/__generated__/types-and-hooks'

export function getTokenSafetyHeaderText(
  safetyLevel: NullUndefined<SafetyLevel>,
  t: TFunction
): string | undefined {
  switch (safetyLevel) {
    case SafetyLevel.MediumWarning:
      return t('Caution')
    case SafetyLevel.StrongWarning:
      return t('Warning')
    case SafetyLevel.Blocked:
      return t('Not available')
  }
}
