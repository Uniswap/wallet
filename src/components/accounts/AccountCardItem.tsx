import React, { useMemo } from 'react'
import { useAppTheme } from 'src/app/hooks'
import Check from 'src/assets/icons/check.svg'
import TripleDots from 'src/assets/icons/triple-dots.svg'
import { AccountIcon } from 'src/components/AccountIcon'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { Flex } from 'src/components/layout'
import { NotificationBadge } from 'src/components/notifications/Badge'
import { Text } from 'src/components/Text'
import { DecimalNumber } from 'src/components/text/DecimalNumber'
import { useENSAvatar } from 'src/features/ens/api'
import { useSelectAddressHasNotifications } from 'src/features/notifications/hooks'
import { ElementName } from 'src/features/telemetry/constants'
import { useDisplayName } from 'src/features/wallet/hooks'
import { iconSizes } from 'src/styles/sizing'
import { formatUSDPrice, NumberType } from 'src/utils/format'

type AccountCardItemProps = {
  address: Address
  isActive?: boolean
  isViewOnly: boolean
  onPress: (address: Address) => void
  onPressEdit: (address: Address) => void
} & PortfolioValueProps

type PortfolioValueProps = {
  isPortfolioValueLoading: boolean
  portfolioValue: number | undefined
}

function PortfolioValue({
  isPortfolioValueLoading,
  portfolioValue,
}: PortfolioValueProps): JSX.Element {
  const isLoading = isPortfolioValueLoading && portfolioValue === undefined

  return (
    <DecimalNumber
      color="textSecondary"
      loading={isLoading}
      number={formatUSDPrice(portfolioValue, NumberType.PortfolioBalance)}
      variant="bodySmall"
    />
  )
}

export function AccountCardItem({
  address,
  isViewOnly,
  isActive,
  isPortfolioValueLoading,
  portfolioValue,
  onPress,
  onPressEdit,
}: AccountCardItemProps): JSX.Element {
  const theme = useAppTheme()
  const displayName = useDisplayName(address)
  const { data: avatar } = useENSAvatar(address)
  const hasNotifications = useSelectAddressHasNotifications(address)

  const icon = useMemo(() => {
    return (
      <AccountIcon
        address={address}
        avatarUri={avatar}
        showViewOnlyBadge={isViewOnly}
        size={iconSizes.icon40}
      />
    )
  }, [address, avatar, isViewOnly])

  return (
    <TouchableArea
      hapticFeedback
      pb="spacing12"
      pt="spacing8"
      px="spacing24"
      onPress={(): void => onPress(address)}>
      <Flex row alignItems="center" testID={`account_item/${address}`}>
        <Flex row shrink alignItems="center">
          <NotificationBadge showIndicator={hasNotifications}>{icon}</NotificationBadge>
          <Flex fill gap="none">
            <Text numberOfLines={1} variant="bodyLarge">
              {displayName?.name}
            </Text>
            <PortfolioValue
              isPortfolioValueLoading={isPortfolioValueLoading}
              portfolioValue={portfolioValue}
            />
          </Flex>
        </Flex>
        <Flex row alignItems="center" gap="none">
          {isActive && (
            <Check
              color={theme.colors.userThemeMagenta}
              height={theme.iconSizes.icon20}
              width={theme.iconSizes.icon20}
            />
          )}
          <TouchableArea
            name={ElementName.Edit}
            pl="spacing12"
            py="spacing16"
            onPress={(): void => onPressEdit(address)}>
            <TripleDots
              color={theme.colors.textTertiary}
              height={iconSizes.icon12}
              strokeLinecap="round"
              strokeWidth="1"
              width={iconSizes.icon16}
            />
          </TouchableArea>
        </Flex>
      </Flex>
    </TouchableArea>
  )
}
