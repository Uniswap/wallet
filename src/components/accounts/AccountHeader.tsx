import { impactAsync, ImpactFeedbackStyle, selectionAsync } from 'expo-haptics'
import React, { useCallback } from 'react'
import { useAppDispatch, useAppSelector, useAppTheme } from 'src/app/hooks'
import { navigate } from 'src/app/navigation/rootNavigation'
import ScanQRWCIcon from 'src/assets/icons/scan-qr-wc.svg'
import SettingsIcon from 'src/assets/icons/settings.svg'
import { AddressDisplay } from 'src/components/AddressDisplay'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { Chevron } from 'src/components/icons/Chevron'
import { Flex } from 'src/components/layout'
import { Box } from 'src/components/layout/Box'
import { ScannerModalState } from 'src/components/QRCodeScanner/constants'
import { openModal } from 'src/features/modals/modalSlice'
import { pushNotification } from 'src/features/notifications/notificationSlice'
import { AppNotificationType } from 'src/features/notifications/types'
import { ElementName, ModalName } from 'src/features/telemetry/constants'
import { selectActiveAccountAddress } from 'src/features/wallet/selectors'
import { removePendingSession } from 'src/features/walletConnect/walletConnectSlice'
import { Screens } from 'src/screens/Screens'
import { iconSizes } from 'src/styles/sizing'
import { setClipboard } from 'src/utils/clipboard'
import { isDevBuild } from 'src/utils/version'

function QRScannerIconButton({ onPress }: { onPress: () => void }): JSX.Element {
  const theme = useAppTheme()

  return (
    <TouchableArea hapticFeedback name={ElementName.WalletConnectScan} onPress={onPress}>
      <ScanQRWCIcon
        color={theme.colors.textTertiary}
        height={theme.iconSizes.icon24}
        opacity="0.8"
        strokeWidth={2}
      />
    </TouchableArea>
  )
}

export function AccountHeader(): JSX.Element {
  const theme = useAppTheme()
  const activeAddress = useAppSelector(selectActiveAccountAddress)
  const dispatch = useAppDispatch()

  const onPressAccountHeader = useCallback(() => {
    dispatch(openModal({ name: ModalName.AccountSwitcher }))
  }, [dispatch])

  const onPressSettings = (): void => {
    navigate(Screens.SettingsStack, { screen: Screens.Settings })
  }

  const onPressScan = useCallback(() => {
    // in case we received a pending session from a previous scan after closing modal
    dispatch(removePendingSession())
    dispatch(
      openModal({ name: ModalName.WalletConnectScan, initialState: ScannerModalState.ScanQr })
    )
  }, [dispatch])

  const onPressCopyAddress = (): void => {
    if (activeAddress) {
      impactAsync()
      setClipboard(activeAddress)
      dispatch(pushNotification({ type: AppNotificationType.Copied }))
    }
  }

  return (
    <Box
      alignItems="center"
      flexDirection="row"
      justifyContent="space-between"
      mt="spacing16"
      testID="account-header">
      <TouchableArea
        hapticFeedback
        alignItems="center"
        flex={1}
        flexDirection="row"
        hapticStyle={ImpactFeedbackStyle.Medium}
        mr="spacing12"
        name={ElementName.Manage}
        testID={ElementName.Manage}
        onLongPress={(): void => {
          onPressCopyAddress()
          if (isDevBuild()) {
            selectionAsync()
            dispatch(openModal({ name: ModalName.Experiments }))
          }
        }}
        onPress={onPressAccountHeader}>
        {activeAddress && (
          <Flex row alignItems="center" gap="spacing4">
            <Flex shrink>
              <AddressDisplay
                hideAddressInSubtitle
                address={activeAddress}
                horizontalGap="spacing8"
                size={iconSizes.icon28}
                variant="subheadLarge"
              />
            </Flex>
            <Chevron
              color={theme.colors.textTertiary}
              direction="s"
              height={iconSizes.icon20}
              width={iconSizes.icon20}
            />
          </Flex>
        )}
      </TouchableArea>
      <Flex alignItems="center" flexDirection="row" gap="spacing16" justifyContent="flex-end">
        <QRScannerIconButton onPress={onPressScan} />
        <TouchableArea hapticFeedback onPress={onPressSettings}>
          <Flex row alignItems="center">
            <SettingsIcon
              color={theme.colors.textTertiary}
              height={theme.iconSizes.icon28}
              opacity="0.8"
              width={theme.iconSizes.icon28}
            />
          </Flex>
        </TouchableArea>
      </Flex>
    </Box>
  )
}
