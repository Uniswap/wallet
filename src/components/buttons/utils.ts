import { ButtonEmphasis, ButtonSize } from 'src/components/buttons/Button'
import { Theme } from 'src/styles/theme'

function getButtonColor(emphasis: ButtonEmphasis): keyof Theme['colors'] {
  switch (emphasis) {
    case ButtonEmphasis.Primary:
      return 'userThemeMagenta'
    case ButtonEmphasis.Secondary:
      return 'background3'
    case ButtonEmphasis.Tertiary:
      return 'none'
    case ButtonEmphasis.Detrimental:
      return 'accentCriticalSoft'
    case ButtonEmphasis.Warning:
      return 'accentWarningSoft'
  }
}

function getButtonTextColor(emphasis: ButtonEmphasis): keyof Theme['colors'] {
  switch (emphasis) {
    case ButtonEmphasis.Primary:
      return 'white'
    case ButtonEmphasis.Secondary:
    case ButtonEmphasis.Tertiary:
      return 'textPrimary'
    case ButtonEmphasis.Detrimental:
      return 'accentCritical'
    case ButtonEmphasis.Warning:
      return 'accentWarning'
  }
}

function getButtonBorderColor(emphasis: ButtonEmphasis): keyof Theme['colors'] {
  switch (emphasis) {
    case ButtonEmphasis.Tertiary:
      return 'backgroundOutline'
    default:
      return 'none'
  }
}

function getButtonTextSizeVariant(size: ButtonSize): keyof Theme['textVariants'] {
  switch (size) {
    case ButtonSize.Large:
      return 'buttonLabelLarge'
    case ButtonSize.Medium:
      return 'buttonLabelMedium'
    case ButtonSize.Small:
      return 'buttonLabelSmall'
  }
}

function getButtonPaddingY(size: ButtonSize): keyof Theme['spacing'] {
  switch (size) {
    case ButtonSize.Large:
      return 'spacing16'
    case ButtonSize.Medium:
      return 'spacing12'
    case ButtonSize.Small:
      return 'spacing8'
  }
}

function getButtonPaddingX(size: ButtonSize): keyof Theme['spacing'] {
  switch (size) {
    case ButtonSize.Large:
      return 'spacing16'
    case ButtonSize.Medium:
      return 'spacing12'
    case ButtonSize.Small:
      return 'spacing8'
  }
}

function getButtonIconPadding(size: ButtonSize): keyof Theme['spacing'] {
  switch (size) {
    case ButtonSize.Large:
      return 'spacing12'
    case ButtonSize.Medium:
      return 'spacing8'
    case ButtonSize.Small:
      return 'spacing4'
  }
}

function getButtonBorderRadius(size: ButtonSize): keyof Theme['borderRadii'] {
  switch (size) {
    case ButtonSize.Large:
      return 'rounded16'
    case ButtonSize.Medium:
      return 'rounded12'
    case ButtonSize.Small:
      return 'rounded8'
  }
}

function getButtonIconSize(size: ButtonSize): keyof Theme['iconSizes'] {
  switch (size) {
    case ButtonSize.Large:
      return 'icon24'
    case ButtonSize.Medium:
      return 'icon20'
    case ButtonSize.Small:
      return 'icon16'
  }
}

export function getButtonProperties(
  emphasis: ButtonEmphasis,
  size: ButtonSize
): {
  backgroundColor: keyof Theme['colors']
  textColor: keyof Theme['colors']
  textVariant: keyof Theme['textVariants']
  paddingX: keyof Theme['spacing']
  paddingY: keyof Theme['spacing']
  borderRadius: keyof Theme['borderRadii']
  borderColor: keyof Theme['colors']
  iconPadding: keyof Theme['spacing']
  iconSize: keyof Theme['iconSizes']
} {
  const backgroundColor = getButtonColor(emphasis)
  const textColor = getButtonTextColor(emphasis)
  const textVariant = getButtonTextSizeVariant(size)
  const paddingX = getButtonPaddingX(size)
  const paddingY = getButtonPaddingY(size)
  const borderColor = getButtonBorderColor(emphasis)
  const borderRadius = getButtonBorderRadius(size)
  const iconPadding = getButtonIconPadding(size)
  const iconSize = getButtonIconSize(size)

  return {
    backgroundColor,
    textColor,
    textVariant,
    paddingX,
    paddingY,
    borderRadius,
    borderColor,
    iconPadding,
    iconSize,
  }
}
