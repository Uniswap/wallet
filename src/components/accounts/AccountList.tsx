import { LinearGradient } from 'expo-linear-gradient'
import { ComponentProps, default as React, useEffect, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet } from 'react-native'
import { ScrollView } from 'react-native-gesture-handler'
import { useAppTheme } from 'src/app/hooks'
import { AccountCardItem } from 'src/components/accounts/AccountCardItem'
import { Box, Flex } from 'src/components/layout'
import { Spacer } from 'src/components/layout/Spacer'
import { Text } from 'src/components/Text'
import { PollingInterval } from 'src/constants/misc'
import { isNonPollingRequestInFlight } from 'src/data/utils'
import { useAccountListQuery } from 'src/data/__generated__/types-and-hooks'
import { Account, AccountType } from 'src/features/wallet/accounts/types'
import { useActiveAccount } from 'src/features/wallet/hooks'
import { spacing } from 'src/styles/sizing'

// Most screens can fit more but this is set conservatively
const MIN_ACCOUNTS_TO_ENABLE_SCROLL = 5

type AccountListProps = Pick<ComponentProps<typeof AccountCardItem>, 'onPress' | 'onPressEdit'> & {
  accounts: Account[]
  isVisible?: boolean
}

type AccountWithPortfolioValue = {
  account: Account
  isPortfolioValueLoading: boolean
  portfolioValue: number | undefined
}

const ViewOnlyHeader = (): JSX.Element => {
  const { t } = useTranslation()
  return (
    <Flex row alignItems="center" borderBottomColor="backgroundOutline">
      <Box flex={1} px="spacing24">
        <Text color="textSecondary" variant="subheadSmall">
          {t('View only')}
        </Text>
      </Box>
    </Flex>
  )
}

export function AccountList({
  accounts,
  onPressEdit,
  onPress,
  isVisible,
}: AccountListProps): JSX.Element {
  const theme = useAppTheme()
  const activeAccount = useActiveAccount()
  const addresses = accounts.map((a) => a.address)

  const { data, networkStatus, refetch, startPolling, stopPolling } = useAccountListQuery({
    variables: { addresses },
    notifyOnNetworkStatusChange: true,
  })

  // Only poll account total values when the account list is visible
  useEffect(() => {
    if (isVisible) {
      refetch()
      startPolling(PollingInterval.Fast)
    } else {
      stopPolling()
    }
  }, [isVisible, refetch, startPolling, stopPolling])

  const isPortfolioValueLoading = isNonPollingRequestInFlight(networkStatus)

  const accountsWithPortfolioValue = useMemo(() => {
    return accounts.map((account, i) => {
      return {
        account,
        isPortfolioValueLoading,
        portfolioValue: data?.portfolios?.[i]?.tokensTotalDenominatedValue?.value,
      } as AccountWithPortfolioValue
    })
  }, [accounts, data, isPortfolioValueLoading])

  const signerAccounts = useMemo(() => {
    return accountsWithPortfolioValue.filter(
      (account) => account.account.type === AccountType.SignerMnemonic
    )
  }, [accountsWithPortfolioValue])

  const viewOnlyAccounts = useMemo(() => {
    return accountsWithPortfolioValue.filter(
      (account) => account.account.type === AccountType.Readonly
    )
  }, [accountsWithPortfolioValue])

  const hasViewOnlyAccounts = viewOnlyAccounts.length > 0

  const renderAccountCardItem = (item: AccountWithPortfolioValue): JSX.Element => (
    <AccountCardItem
      key={item.account.address}
      address={item.account.address}
      isActive={!!activeAccount && activeAccount.address === item.account.address}
      isPortfolioValueLoading={item.isPortfolioValueLoading}
      isViewOnly={item.account.type === AccountType.Readonly}
      portfolioValue={item.portfolioValue}
      onPress={onPress}
      onPressEdit={onPressEdit}
    />
  )

  return (
    <Box flexShrink={1} position="relative">
      {/* TODO: [MOB-3938] attempt to switch gradients to react-native-svg#LinearGradient and avoid new clear color */}
      <LinearGradient
        colors={[theme.colors.clearBackground1Backdrop, theme.colors.background1]}
        end={{ x: 0, y: 0 }}
        start={{ x: 0, y: 1 }}
        style={ListSheet.topGradient}
      />
      <ScrollView
        bounces={false}
        scrollEnabled={accountsWithPortfolioValue.length >= MIN_ACCOUNTS_TO_ENABLE_SCROLL}
        showsVerticalScrollIndicator={false}>
        {signerAccounts.map(renderAccountCardItem)}
        {hasViewOnlyAccounts && (
          <>
            <Spacer height={theme.spacing.spacing8} />
            <ViewOnlyHeader />
            {viewOnlyAccounts.map(renderAccountCardItem)}
          </>
        )}
      </ScrollView>
      <LinearGradient
        colors={[theme.colors.clearBackground1Backdrop, theme.colors.background1]}
        end={{ x: 0, y: 1 }}
        start={{ x: 0, y: 0 }}
        style={ListSheet.bottomGradient}
      />
    </Box>
  )
}

const ListSheet = StyleSheet.create({
  bottomGradient: {
    bottom: 0,
    height: spacing.spacing16,
    left: 0,
    position: 'absolute',
    width: '100%',
  },
  topGradient: {
    height: spacing.spacing16,
    left: 0,
    position: 'absolute',
    top: 0,
    width: '100%',
    zIndex: 1,
  },
})
