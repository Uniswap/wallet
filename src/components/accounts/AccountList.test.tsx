import { MockedResponse } from '@apollo/client/testing'
import { fireEvent } from '@testing-library/react-native'
import React from 'react'
import { AccountList } from 'src/components/accounts/AccountList'
import { ON_PRESS_EVENT_PAYLOAD } from 'src/components/buttons/TouchableArea.test'
import { AccountListDocument, AccountListQuery } from 'src/data/__generated__/types-and-hooks'
import { account } from 'src/test/fixtures'
import { Portfolios } from 'src/test/gqlFixtures'
import { render, screen } from 'src/test/test-utils'
import { formatUSDPrice, NumberType } from 'src/utils/format'

jest.mock('@react-navigation/drawer', () => ({ useDrawerStatus: jest.fn(() => 'unknown') }))

const mock: MockedResponse<AccountListQuery> = {
  request: {
    query: AccountListDocument,
    variables: {
      addresses: [account.address],
    },
  },
  result: {
    data: {
      portfolios: Portfolios,
    },
  },
}

describe(AccountList, () => {
  it('renders without error', async () => {
    const tree = render(
      <AccountList accounts={[account]} onPress={jest.fn()} onPressEdit={jest.fn()} />,
      { mocks: [mock] }
    )

    expect(
      await screen.findByText(
        formatUSDPrice(
          Portfolios[0].tokensTotalDenominatedValue?.value,
          NumberType.PortfolioBalance
        )
      )
    ).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('handles press on card items', async () => {
    const onPressSpy = jest.fn()
    render(<AccountList accounts={[account]} onPress={onPressSpy} onPressEdit={jest.fn()} />, {
      mocks: [mock],
    })
    // go to success state
    expect(
      await screen.findByText(
        formatUSDPrice(
          Portfolios[0].tokensTotalDenominatedValue?.value,
          NumberType.PortfolioBalance
        )
      )
    ).toBeDefined()
    fireEvent.press(screen.getByTestId(`account_item/${account.address}`), ON_PRESS_EVENT_PAYLOAD)

    expect(onPressSpy).toHaveBeenCalledTimes(1)
  })
})
