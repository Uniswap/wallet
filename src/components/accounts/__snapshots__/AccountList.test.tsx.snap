// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountList renders without error 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "flexShrink": 1,
        "position": "relative",
      },
    ]
  }
>
  ExpoLinearGradient
  <RCTScrollView
    bounces={false}
    collapsable={false}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    scrollEnabled={false}
    showsVerticalScrollIndicator={false}
    waitFor={
      Array [
        Object {
          "current": null,
        },
      ]
    }
  >
    <View>
      <ContextMenu
        actions={
          Array [
            Object {
              "systemIcon": "doc.on.doc",
              "title": "Copy wallet address",
            },
            Object {
              "systemIcon": "gearshape",
              "title": "Wallet settings",
            },
          ]
        }
        onPress={[Function]}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 5,
              "left": 5,
              "right": 5,
              "top": 5,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "display": "flex",
              "opacity": 1,
              "paddingBottom": 12,
              "paddingHorizontal": 24,
              "paddingTop": 8,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                },
              ]
            }
            testID="account_item/0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "flexShrink": 1,
                    "justifyContent": "flex-start",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "display": "flex",
                      "height": "100%",
                      "position": "relative",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderColor": "transparent",
                        "borderRadius": 999999,
                        "borderWidth": 0,
                        "display": "flex",
                        "position": "relative",
                      },
                      Object {
                        "padding": 0,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "display": "flex",
                          "height": 40,
                          "width": 40,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        clip={true}
                        mask={
                          <UniconMask
                            attributeData={
                              Object {
                                "0": "#5CFE9D",
                                "1": "#B8C3B7",
                                "2": Array [
                                  Object {
                                    "path": "M34 20C34 28.8366 26.8366 36 18 36C9.16344 36 2 28.8366 2 20L2 -3.8147e-06L34 -1.01717e-06V20Z",
                                  },
                                ],
                                "3": Array [
                                  Object {
                                    "fillType": "evenOdd",
                                    "path": "M16 8C12.2937 6.62854 9.37146 3.70632 8 0C6.62854 3.70632 3.70632 6.62854 0 8C3.70632 9.37146 6.62854 12.2937 8 16C9.37146 12.2937 12.2937 9.37146 16 8ZM8 11C9.65685 11 11 9.65685 11 8C11 6.34315 9.65685 5 8 5C6.34315 5 5 6.34315 5 8C5 9.65685 6.34315 11 8 11Z",
                                  },
                                ],
                              }
                            }
                            size={40}
                          />
                        }
                      >
                        <View
                          transform={
                            Array [
                              Object {
                                "scale": 1.0810810810810811,
                              },
                            ]
                          }
                        >
                          <View
                            height={37}
                            width={37}
                            x={0}
                            y={0}
                          >
                            <View
                              colors={
                                Array [
                                  "#5CFE9D",
                                  "#B8C3B7",
                                ]
                              }
                              end={null}
                              start={null}
                            />
                          </View>
                          <View
                            color="#EDE590"
                            cx={18.5}
                            cy={-13.36111111111111}
                            r={30.833333333333336}
                          >
                            <View
                              blur={15}
                            />
                          </View>
                        </View>
                      </View>
                      <View
                        clip={true}
                        mask={
                          <UniconMask
                            attributeData={
                              Object {
                                "0": "#5CFE9D",
                                "1": "#B8C3B7",
                                "2": Array [
                                  Object {
                                    "path": "M34 20C34 28.8366 26.8366 36 18 36C9.16344 36 2 28.8366 2 20L2 -3.8147e-06L34 -1.01717e-06V20Z",
                                  },
                                ],
                                "3": Array [
                                  Object {
                                    "fillType": "evenOdd",
                                    "path": "M16 8C12.2937 6.62854 9.37146 3.70632 8 0C6.62854 3.70632 3.70632 6.62854 0 8C3.70632 9.37146 6.62854 12.2937 8 16C9.37146 12.2937 12.2937 9.37146 16 8ZM8 11C9.65685 11 11 9.65685 11 8C11 6.34315 9.65685 5 8 5C6.34315 5 5 6.34315 5 8C5 9.65685 6.34315 11 8 11Z",
                                  },
                                ],
                              }
                            }
                            overlay={true}
                            size={40}
                          />
                        }
                      >
                        <View
                          color="#000000"
                          height={37}
                          opacity={0.08}
                          width={37}
                          x={0}
                          y={0}
                        />
                      </View>
                    </View>
                  </View>
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "display": "flex",
                      "flexGrow": 0,
                      "flexShrink": 0,
                      "height": 16,
                      "width": 16,
                    },
                  ]
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "stretch",
                      "display": "flex",
                      "flex": 1,
                      "flexDirection": "column",
                      "justifyContent": "flex-start",
                    },
                  ]
                }
              >
                <Text
                  allowFontScaling={true}
                  maxFontSizeMultiplier={1.4}
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#0E1524",
                        "display": "flex",
                        "fontFamily": "Inter-Medium",
                        "fontSize": 20,
                        "lineHeight": 24,
                      },
                    ]
                  }
                >
                  0xâ€‹f39F...2266
                </Text>
                <View
                  style={
                    Array [
                      Object {
                        "display": "flex",
                        "flexGrow": 0,
                        "flexShrink": 0,
                        "height": 0,
                      },
                    ]
                  }
                />
                <Text
                  allowFontScaling={true}
                  maxFontSizeMultiplier={1.4}
                  style={
                    Array [
                      Object {
                        "color": "#888FAB",
                        "display": "flex",
                        "fontFamily": "Inter-Medium",
                        "fontSize": 15,
                        "lineHeight": 20,
                      },
                    ]
                  }
                >
                  $55.00
                </Text>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "height": 16,
                    "width": 16,
                  },
                ]
              }
            />
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  },
                ]
              }
            />
          </View>
        </View>
      </ContextMenu>
    </View>
  </RCTScrollView>
  ExpoLinearGradient
</View>
`;
